File: src/bin/allelua/lua/table.rs
use mlua::chunk;

pub fn load_table(lua: &'static mlua::Lua) -> mlua::Result<()> {
    let is_empty = lua.create_function(|_lua, t: mlua::Table| Ok(t.is_empty()))?;

    lua.load(chunk! {
        local table = require("table")
        local M = table

        // LuaJIT extensions.
        M.new = require("table.new")
        M.clear = require("table.clear")

        // Freeze table.
        M.frozen_table_key = "__freeze"
        M.frozen_table_mt = {
          __index = function(t, k)
            return t[M.frozen_table_key][k]
          end,
          __newindex = function()
            error("attempt to update a frozen table", 2)
          end,
          __len       = function(t) return #t[M.frozen_table_key]          end,
          __tostring  = function(t, opts) return tostring(t[M.frozen_table_key], opts) end,
          __pairs     = function(t) return pairs(t[M.frozen_table_key])    end,
          __ipairs    = function(t) return ipairs(t[M.frozen_table_key])   end,
          __iter      = function(t) return iter(t[M.frozen_table_key])     end,
          __metatable = false, // protect metatable
        }
        M.freeze = function(t)
          assert(type(t) == "table", "invalid input: " .. tostring(t) .. " is not a table")
          if t[M.frozen_table_key] ~= nil then return t end
          local proxy = { [M.frozen_table_key] = t }
          setmetatable(proxy, M.frozen_table_mt)
          return proxy
        end
        M.is_frozen = function(t)
            return t[M.frozen_table_key] ~= nil
        end

        M.is_empty = $is_empty

        M.push = function(t, ...)
            local args = {...}
            for _, v in ipairs(args) do
                t[#t + 1] = v
            end
            return t
        end

        M.map = function(t, map_fn)
            local result = {}
            for k, v in pairs(t) do
                local new_k, new_v = map_fn(k, v)
                if new_v == nil then
                    new_v = new_k
                    new_k = k
                end
                t[new_k] = new_v
            end
            setmetatable(result, getmetatable(t))
            return result
        end

        M.clone = function(t)
            local t_mt = getmetatable(t)
            local can_pairs = t_mt and t_mt.__pairs
            if not can_pairs then return t end

            local clone = {}
            for k, v in pairs(t) do
                clone[k] = v
            end

            return setmetatable(clone, getmetatable(t))
        end

        M.deep_clone = function(t)
            local t_mt = getmetatable(t)
            local can_pairs = t_mt and t_mt.__pairs
            if not can_pairs then return t end

            local clone = {}
            for k, v in pairs(t) do
                clone[M.deep_clone(k)] = M.deep_clone(v)
            end

            return setmetatable(clone, getmetatable(t))
        end

        M.deep_eq = function(a, b)
            if a == b then return true end

            // If either value is not a table, they're not equal (since a ~= b)
            if type(a) ~= "table" or type(b) ~= "table" then
                return false
            end

            // Fast is_empty check.
            if M.is_empty(a) ~= M.is_empty(b) then return false end

            // Check for cycles
            seen = seen or {}
            if seen[a] and seen[a][b] then
                return true // We've seen this pair before, consider them equal to avoid infinite recursion
            end
            seen[a] = seen[a] or {}
            seen[a][b] = true
            seen[b] = seen[b] or {}
            seen[b][a] = true

            // Check if all keys in 'a' exist in 'b' and have the same values
            for k, v in pairs(a) do
                if not M.deep_eq(v, b[k], seen) then
                    return false
                end
            end

            // Check if 'b' has any keys that 'a' doesn't have
            for k in pairs(b) do
                if a[k] == nil then
                    return false
                end
            end

            return true
        end

        M.collect_map = function(map_fn)
            return function(iterator, state, initial_value)
                local result = {}
                for k, v in iterator, state, initial_value do
                    table.insert(result, map_fn(k, v))
                end

                return result
            end
        end

        M.collect = M.collect_map(M.pack)
        M.collect_entries = M.collect_map(function(k, v) return { k, v } end)
        M.collect_keys = M.collect_map(function (k, v) return k end)
        M.collect_values = M.collect_map(function (k, v) return v end)

        M.keys = function(table) return M.collect_keys(pairs(table)) end
        M.values = function(table) return M.collect_values(pairs(table)) end
        M.ivalues = function(table) return M.collect_values(ipairs(table)) end
        M.entries = function(table) return M.collect_entries(pairs(table)) end
        M.from_entries = function(entries)
            local result = {}
            for _, entry in ipairs(entries) do
                local k, v = table.unpack(entry)
                result[k] = v
            end

            return result
        end
    })
    .exec()?;

    Ok(())
}

File: src/bin/allelua/lua/table/table.lua
return function(is_empty)
	local table = require("table")
	local math = require("math")
	local M = table

	local function range_index(t, i, j)
		local len = M.getn(t)
		i = i or 1
		j = j or len

		if i < 0 then i = len + 1 + i end
		if j < 0 then j = len + 1 + j end

		if i <= 0 then i = 1 end
		if j > len then j = len end

		return i, j, len
	end

	-- LuaJIT extensions.
	M.new = require("table.new")
	M.clear = require("table.clear")

	M.for_each = M.foreach
	M.foreach = nil
	M.ifor_each = M.foreachi
	M.foreachi = nil

	local sort = M.sort
	M.sort = function(t, comp)
		sort(t, comp)
		return t
	end

	M.dedup = function(t)
		local n = 0
		local prev = nil
		local len = M.getn(t)

		for i = len, 1, -1 do
			if t[i] ~= prev then
				if n >= 2 then M.delete(t, i + 1, i + n - 1) end
				prev = t[i]
				n = 1
			else
				n = n + 1
			end
		end

		if n >= 2 then M.delete(t, 1, n - 1) end

		return t
	end

	M.is_empty = is_empty

	M.reverse = function(t)
		local len = M.getn(t)
		for i = 1, math.floor(len / 2) do
			local a = t[i]
			local b = t[len - i + 1]
			t[i] = b
			t[len - i + 1] = a
		end
		return t
	end

	M.binary_search = function(t, elem)
		local left = 1
		local right = M.getn(t)

		while left <= right do
			local mid = left + math.round((right - left) / 2)

			if t[mid] == elem then
				return mid
			elseif t[mid] < elem then
				left = mid + 1
			else
				right = mid - 1
			end
		end

		return nil
	end

	M.linear_search = function(t, elem)
		for i, e in ipairs(t) do
			if e == elem then return i end
		end
		return nil
	end

	M.delete = function(t, i, j)
		local len = nil
		i, j, len = range_index(t, i, j or i)

		if i == j then
			local v = t[i]
			M.remove(t, i)
			return { v }
		end
		local removed = {}

		M.move(t, i, j, 1, removed)
		M.move(t, j + 1, len, i)
		M.move({}, 1, #removed, (len + 1) - #removed, t)
		return removed
	end

	M.push = function(t, ...)
		local args = { ... }
		for _, v in ipairs(args) do
			M.insert(t, v)
		end
		return M.getn(t)
	end

	M.pop = function(t)
		return M.remove(t)
	end

	M.unshift = function(t, ...)
		local args = { ... }
		for _, v in ipairs(args) do
			M.insert(t, 1, v)
		end
		return M.getn(t)
	end

	M.shift = function(t)
		return M.remove(t, 1)
	end

	M.indexof = function(t, elem, start)
		start = start or 1
		local i = start
		while true do
			if t[i] == elem then return i end
			if t[i] == nil then break end
			i = i + 1
		end
		return -1
	end

	M.slice = function(t, i, j)
		i, j = range_index(t, i, j)
		if j < i then return {} end
		local slice = M.new(j + 1 - i, 0)
		for index = i, j do
			table.push(slice, t[index])
		end

		return slice
	end

	M.flat = function(t, depth, result)
		depth = depth or 1
		result = result or {}

		for _, v in ipairs(t) do
			if depth > 0 and type(v) == "table" then
				result = M.flat(v, depth - 1, result)
			else
				M.insert(result, v)
			end
		end

		return result
	end

	M.map = function(t, map_fn)
		local result = {}
		for k, v in pairs(t) do
			local new_k, new_v = map_fn(k, v)
			if new_v == nil then
				new_v = new_k
				new_k = k
			end
			result[new_k] = new_v
		end
		setmetatable(result, getmetatable(t))
		return result
	end

	M.map_keys = function(t, map_fn)
		return M.map(t, function(k, v)
			k = map_fn(k)
			return k, v
		end)
	end
	M.map_values = function(t, map_fn)
		return M.map(t, function(k, v)
			v = map_fn(v)
			return k, v
		end)
	end

	M.deep_eq = function(a, b, seen)
		if a == b then return true end

		-- If either value is not a table, they're not equal (since a ~= b)
		if rawtype(a) ~= "table" or rawtype(b) ~= "table" then return false end

		-- We can't use M.is_empty for fast checks has table may have __index
		-- metamethod.

		-- Check for cycles
		seen = seen or {}
		if seen[a] and seen[a][b] then
			return true -- We've seen this pair before, consider them equal to avoid infinite recursion
		end
		seen[a] = seen[a] or {}
		seen[a][b] = true
		seen[b] = seen[b] or {}
		seen[b][a] = true

		-- Check if all keys in 'a' exist in 'b' and have the same values
		for k, v in pairs(a) do
			if not M.deep_eq(v, b[k], seen) then return false end
		end

		-- Check if 'b' has any keys that 'a' doesn't have
		for k in pairs(b) do
			if a[k] == nil then return false end
		end

		return true
	end

	M.icollect = function(iter, state, initial)
		local t = {}

		for i, v in iter, state, initial do
			if i and not v then v = i end
			table.push(t, v)
		end

		return t
	end

	M.collect = function(iter, state, initial)
		local t = {}

		for k, v in iter, state, initial do
			if k and not v then v = k end
			t[k] = v
		end

		return t
	end
end
